compiler/rustc_ast_passes/src/ast_validation.rs

```diff
@@ -213,6 +229,85 @@ impl<'a> AstValidator<'a> {
         err.emit();
     }
 
+    fn check_anonymous_field(&self, field: &FieldDef) {
+        let FieldDef { ty, ident, .. } = field;
+        match &ty.kind {
+            TyKind::AnonymousStruct(..) | TyKind::AnonymousUnion(..) => {
+                if let Some(Ident { name: kw::Underscore, .. }) = ident {
+                    // Empty because we want to error only if Ident != `_`
+                } else {
+                    let struct_kind = if matches!(ty.kind, TyKind::AnonymousStruct(..)) {
+                        "struct"
+                    } else {
+                        "union"
+                    };
+                    let msg =
+                        format!("anonymous {} not allowed with a named identifier", struct_kind);
+                    self.err_handler()
+                        .struct_span_err(field.span, msg.as_str())
+                        // There are no `None` idents on a struct field def, so it should be safe to unwrap
+                        .span_suggestion(
+                            ident.unwrap().span,
+                            "replace with",
+                            "_".to_string(),
+                            Applicability::MachineApplicable,
+                        )
+                        .emit();
+                }
+            }
+            TyKind::Path(..) => {
+                // If the anonymous field contains a Path as type, we can't determine if the path is a valid
+                // struct or union, so we skip the check
+            }
+            _ => {
+                if let Some(Ident { name: kw::Underscore, .. }) = ident {
+                    let msg = "only anonymous structs and unions are allowed";
```

Comment:
```suggestion
                    let msg = "unnamed fields can only have struct or union types";
```
Non-anonymous structs and unions are allowed too.